python_modules/dagster/dagster/core/definitions/assets_info.py

```diff
@@ -64,6 +67,188 @@ def __new__(
+    curr_node_handle: The current node handle being traversed.
+    graph_output_name: Name of the node output being traversed. Only used if the current node is a
+        graph to trace the op that generates this output.
+    """
+
+    if curr_node_handle in dep_node_handles_by_node:
+        return dep_node_handles_by_node[curr_node_handle]
+
+    dependency_node_handles: List[
+        NodeHandle
+    ] = []  # first node in list is node that outputs the asset
+    if curr_node_handle not in outputs_by_graph_handle:
+        dependency_node_handles.append(curr_node_handle)
+    else:  # is graph
+        output_name, output_node = outputs_by_graph_handle[curr_node_handle][graph_output_name]
+        dependency_node_handles.extend(
+            _get_dependency_node_handles(
+                non_asset_inputs_by_node_handle,
+                outputs_by_graph_handle,
+                dep_node_handles_by_node,
+                output_node,
+                output_name,
+            )
+        )
+    for node_output_name, node_handle in non_asset_inputs_by_node_handle[curr_node_handle]:
+        dependency_node_handles.extend(
+            _get_dependency_node_handles(
+                non_asset_inputs_by_node_handle,
+                outputs_by_graph_handle,
+                dep_node_handles_by_node,
+                node_handle,
+                node_output_name,
+            )
+        )
+
+    if curr_node_handle not in outputs_by_graph_handle:
+        dep_node_handles_by_node[curr_node_handle] = dependency_node_handles
+
+    return dependency_node_handles
+
+
+def _asset_key_to_dep_node_handles(
+    graph_def: GraphDefinition, assets_defs_by_node_handle: Mapping[NodeHandle, "AssetsDefinition"]
+) -> Mapping[AssetKey, Set[NodeHandle]]:
+    """
+    For each asset in assets_defs_by_node_handle, returns all the op handles within the asset's node
+    that are upstream dependencies of the asset.
+    """
+    # A mapping of all node handles to all upstream node handles
+    # that are not assets. Each key is a node handle with tuple value (node_output_name, output_node)
+    non_asset_inputs_by_node_handle: Dict[NodeHandle, Sequence[Tuple[str, NodeHandle]]] = {}
+
+    # A mapping of every graph node handle to a dictionary with each out
+    # name as a key and a tuple containing (op output name, op node handle) as the value
+    outputs_by_graph_handle: Dict[NodeHandle, Dict[str, Tuple[str, NodeHandle]]] = {}
+    _build_graph_dependencies(
+        graph_def=graph_def,
+        parent_handle=None,
+        outputs_by_graph_handle=outputs_by_graph_handle,
+        non_asset_inputs_by_node_handle=non_asset_inputs_by_node_handle,
+        assets_defs_by_node_handle=assets_defs_by_node_handle,
+    )
+
+    dep_nodes_by_asset_key: Dict[AssetKey, List[NodeHandle]] = {}
+
+    for node_handle, assets_defs in assets_defs_by_node_handle.items():
+        dep_node_handles_by_node: Dict[
```

Comment:
Could this be taken out of the for loop? Seems like we could avoid redoing work by sharing these across all the assets, given that it's all the same graph?

Reply:
I think this dictionary should stay here--each `node_handle` in this outer loop is a top level op/graph in the asset job, and there won't be any shared node handles that are nested in different top level `node_handle`s. So this dictionary exists to memoize traversals for every single asset output for the given top level `node_handle`.